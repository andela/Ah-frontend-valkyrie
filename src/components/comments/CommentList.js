import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { fetchComments } from "./actions/actions";
import SingleComment from "./SingleComment";

export class Comment extends Component {
  constructor(props) {
    super(props);
    this.state = {
      comments: [],
    };
  }

  componentDidMount() {
    const { articleSlug } = this.props;
    this.props.fetchComments(articleSlug);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ comments: nextProps.comments });
  }

  render() {
    const { comments } = this.state;
    const { articleSlug } = this.props;
    return (
      <Fragment>
        <div className="comments mt-5">
          <h6>Comments</h6>
          <hr />
          {comments.map(comment => (
            <SingleComment
              key={comment.id}
              comment={comment}
              articleSlug={articleSlug}
              commentBody={comment.body}
              commentId={comment.id}
              username={comment.author.username}
              createdOn={comment.createdAt}
              commentHistory={comment.comment_history}
            />
          ))}
        </div>
      </Fragment>
    );
  }
}

Comment.propTypes = {
  fetchComments: PropTypes.func.isRequired,
  articleSlug: PropTypes.string.isRequired,
  comments: PropTypes.instanceOf(Array).isRequired,
};

const mapStateToProps = state => ({
  comments: state.comments.comments,
});

export default connect(
  mapStateToProps,
  { fetchComments },
)(Comment);
