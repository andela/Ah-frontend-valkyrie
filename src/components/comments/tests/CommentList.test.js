import React from 'react';
import { shallow } from 'enzyme';
import {Comment} from '../CommentList';

const mockCommentOne = {
    id: 78,
    body: 'fake body',
    author: {
        username: 'tester',
    },
    createdAt: 'Mon-25-Feb-2019',
};

const mockCommentTwo = {
    id: 94,
    body: 'fake body two',
    author: {
        username: 'tester-two',
    },
    createdAt: 'Mon-25-Feb-2019',
};

const mockCommentThree = {
    id: 103,
    body: 'fake body three',
    author: {
        username: 'tester-three',
    },
    createdAt: 'Mon-25-Feb-2019',
};

const props={
    comments:[{}],
    fetchComments:jest.fn(),
    articleSlug:""
}

const articleSlug = 'test-article-slug';
const commentsList = [ mockCommentOne, mockCommentTwo, mockCommentThree ];

describe('<Comment />', () => {
    let mockComponent;
    beforeEach(() => {
        mockComponent = shallow( <Comment {...props}/> );
        mockComponent.defaultProps = { articleSlug };
    })

    // component.setState( { comments: commentsList } );
    it('renders as expected ', () => {
      expect(mockComponent).toMatchSnapshot();
    });
    it('should receive next props',()=>{
        mockComponent.instance().componentWillReceiveProps({comments: [mockCommentOne]});
        // mockComponent.setProps({comments: [mockCommentOne]});
        expect(mockComponent.state('comments')).toEqual([mockCommentOne]);
    }),

    it('should mount component',()=>{
        mockComponent.instance().componentDidMount();
        expect(props.fetchComments).toBeCalled();
    })
  });