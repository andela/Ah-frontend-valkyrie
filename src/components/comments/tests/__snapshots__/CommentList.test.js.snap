// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> renders as expected  1`] = `
ShallowWrapper {
  "defaultProps": Object {
    "articleSlug": "test-article-slug",
  },
}
`;
